python/test/unit/fem/test_assemble_submesh.py

```diff
@@ -100,3 +99,69 @@ def test_submesh_facet_assembly(n, k, space, ghost_mode):
     assert A_submesh.squared_norm() == pytest.approx(A_square_mesh.squared_norm(), rel=1.0e-5, abs=1.0e-5)
     assert b_submesh.norm() == pytest.approx(b_square_mesh.norm())
     assert np.isclose(s_submesh, s_square_mesh)
+
+
+@pytest.mark.parametrize("n", [2, 6])
+@pytest.mark.parametrize("k", [1, 4])
+@pytest.mark.parametrize("space", ["Lagrange", "Discontinuous Lagrange"])
+@pytest.mark.parametrize("ghost_mode", [GhostMode.none, GhostMode.shared_facet])
+def test_mixed_dom_codim_0(n, k, space, ghost_mode):
+    """Test assembling a form where the trial and test functions
+    are defined on different meshes"""
+
+    msh = create_rectangle(MPI.COMM_WORLD, ((0.0, 0.0), (2.0, 1.0)), (2 * n, n), ghost_mode=ghost_mode)
+
+    # Locate cells in left half of mesh and create mesh tags
+    tdim = msh.topology.dim
+    cells = locate_entities(msh, tdim, lambda x: x[0] <= 1.0)
+    perm = np.argsort(cells)
+    tag = 1
+    values = np.full_like(cells, tag, dtype=np.intc)
+    ct = meshtags(msh, tdim, cells[perm], values[perm])
+
+    # Create integration measure on the mesh
+    dx_msh = ufl.Measure("dx", domain=msh, subdomain_data=ct)
+
+    # Create a submesh of the left half of the mesh
+    smsh, smhs_to_msh = create_submesh(msh, tdim, cells)[:2]
+
+    # Define function spaces over the mesh and submesh
+    V_msh = fem.functionspace(msh, (space, k))
+    V_smsh = fem.functionspace(smsh, (space, k))
+
+    # Trial and test functions on the mesh
+    u, v = ufl.TrialFunction(V_msh), ufl.TestFunction(V_msh)
+
+    # Test function on the submesh
+    w = ufl.TestFunction(V_smsh)
+
+    def ufl_form(u, v, dx):
+        return ufl.inner(u, v) * dx
+
+    # Define form to compare to and assemble
+    a = fem.form(ufl_form(u, v, dx_msh(tag)))
+    A = assemble_matrix(a)
+    A.assemble()
+
+    # Assemble a mixed domain form, taking smsh to be the integration domain
+    # Entity maps must take cells in smsh (the integration domain mesh) to
+    # cells in msh
+    entity_maps = {msh._cpp_object: np.array(smhs_to_msh, dtype=np.int32)}
+    a_0 = fem.form(ufl_form(u, w, ufl.dx(smsh)), entity_maps=entity_maps)
+    A_0 = assemble_matrix(a_0)
+    A_0.assemble()
+    assert np.isclose(A_0.norm(), A.norm())
+
+    # Now assemble a mixed domain form using msh as integration domain
+    # Entity maps must take cells in msh (the integration domain mesh) to
+    # cells in smsh
+    cell_imap = msh.topology.index_map(tdim)
+    num_cells = cell_imap.size_local + cell_imap.num_ghosts
+    msh_to_smsh = np.full(num_cells, -1)
+    msh_to_smsh[smhs_to_msh] = np.arange(len(smhs_to_msh))
+    entity_maps = {smsh._cpp_object: np.array(msh_to_smsh, dtype=np.int32)}
```

Comment:
Do we check what happens if we send in a -1 in the map, and we try to use it in the kernels?
This should probably raise an error saying invalid map (maybe an assert).
Otherwise it would be hard to debug custom maps.